core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -253,24 +265,30 @@ class ReplicaFetcherThread(name: String,
     val partitionsWithError = mutable.Set[TopicPartition]()
 
     fetchedEpochs.foreach { case (tp, epochOffset) =>
-      val replica = replicaMgr.getReplica(tp).get
-
-      if (epochOffset.hasError) {
-        info(s"Retrying leaderEpoch request for partition ${replica.topicPartition} as the leader reported an error: ${epochOffset.error}")
-        partitionsWithError += tp
-      } else {
-        val truncationOffset =
-          if (epochOffset.endOffset == UNDEFINED_EPOCH_OFFSET)
-            highWatermark(replica, epochOffset)
-          else if (epochOffset.endOffset >= replica.logEndOffset.messageOffset)
-            logEndOffset(replica, epochOffset)
-          else
-            epochOffset.endOffset
-
-        truncationPoints.put(tp, truncationOffset)
+      try {
+        val replica = replicaMgr.getReplicaOrException(tp)
+
+        if (epochOffset.hasError) {
+          info(s"Retrying leaderEpoch request for partition ${replica.topicPartition} as the leader reported an error: ${epochOffset.error}")
+          partitionsWithError += tp
+        } else {
+          val truncationOffset =
+            if (epochOffset.endOffset == UNDEFINED_EPOCH_OFFSET)
+              highWatermark(replica, epochOffset)
+            else if (epochOffset.endOffset >= replica.logEndOffset.messageOffset)
+              logEndOffset(replica, epochOffset)
+            else
+              epochOffset.endOffset
+
+          replicaMgr.logManager.truncateTo(Map(tp -> truncationOffset))
+          truncationPoints.put(tp, truncationOffset)
+        }
+      } catch {
+        case e: KafkaStorageException =>
+          info(s"Failed to truncate $tp", e)
+          partitionsWithError += tp
```

Comment:
Similarly here, it seems it will be cleaner if we return a set of error partitions to the caller and let AbstractFetcherThread call handlePartitionsWithErrors(). It seems the current the code already directly calls delayPartitions. But it should really call handlePartitionsWithErrors() instead.

Reply:
Yeah I have updated the patch to replace `delayPartitions` with `handlePartitionsWithErrors`. But I am not sure if we should change the return value of these methods to be 2-tuple in order to return error partitions to the caller. It seems that handlePartitionsWithErrors() is idempotent when invoked multiple times on the same partition.